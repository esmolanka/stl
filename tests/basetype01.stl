
type API =
  { truncate+ : Float -> Int
  , fromInt+  : Int -> Float
  , showInt+  : Int -> String
  }

#check forall i f s.
       { truncate+ : f -> i
       , fromInt+  : i -> f
       , showInt+  : i -> s
       }
       <: API

type Array a = exists (n : Nat).
  { size+     : Nat n
  , elements+ : a[n]
  }[n]

#eval Array
#eval Array (<Int : Int, Float : Float>)
