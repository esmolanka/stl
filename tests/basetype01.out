tests/basetype01.types:8:1-13:14:
  Subtype:
    ∀ i. ∀ f. ∀ s. ∃ (ρ : Row).
      {truncate : f -> i, fromInt : i -> f, showInt : i -> s | ρ}
  Supertype:
    ∃ (ρ : Row).
      { truncate : Float -> Int
      , fromInt : Int -> Float
      , showInt : Int -> String
      | ρ
      }
  Result:
    OK
tests/basetype01.types:20:1-20:12:
  Normalised:
    λ a. ∃ (n : Nat). (∃ (ρ : Row). {size : Nat n, elements : a[n] | ρ})[n]
  Kind:
    Type -> Type
tests/basetype01.types:21:1-21:40:
  Normalised:
    ∃ (n : Nat).
      (∃ (ρ : Row).
        { size : Nat n
        , elements :
            (∀ (ρ : Row). ∃ (ω : #). ∃ (ω : #).
              <Int^ω : Int| Float^ω/1 : Float | ... ρ>)[n]
        | ρ
        })[n]
  Kind:
    Type